### PR TITLE
non-reflective abstractions for Java codegen Contracts

### LINKED ISSUE #13471: [FEATURE REQUEST] Make Java codegen more OO friendly by adding contract methods to empty contract Interface
I confirm that, if this is a feature request that has security implications, I already contacted security@digitalasset.com and followed the [responsible disclosure policy](https://digitalasset.com/security/).

I confirm that this is not a question or a request for technical support by the community, for which the [Daml forum](https://discuss.daml.com/) is available.

### What is the problem you want to solve?

When generating the [Java bindings](https://github.com/digital-asset/daml/blob/7000875e6287da3d221896ba8d16f68f9e83d7da/language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ContractClass.scala) all generated Templates seem to contain a static inner Contract class which itself implements an [empty contract interface](https://github.com/digital-asset/daml/blob/d2e2c21684e2d9eade40f74990208125ad4b4ccb/language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/Contract.java). These static inner Contract classes are all generated with the methods fromIdAndRecord, fromIdAndRecord (deprecated) and fromCreatedEvent

### What is the solution you would propose?

The addition of the methods fromIdAndRecord, fromIdAndRecord (deprecated) and fromCreatedEvent to the Contract interface would allow for derivative code to leverage generics better using the interface.

### Describe alternatives you've considered

Similar consideration could also be given to the abstract Template class itself by moving the static Identifier variable onto the parent Template class (or exposing a getter for it)

### Additional context

None for now

### PR DESCRIPTION
**Pull request by S11001001**
Fixes #13471. `readActiveContractKeys` and `readActiveContractPayloads` are simple examples of functions that would not have been possible to write before.

* well-typed parents for generated `ContractId` and `Contract`
* move a bunch of things to these classes instead of code-generating them
* static `COMPANION` as a now-and-future home for well-typed abstractions over things that are currently just static methods

Eta-expanding static methods into functions works for many use cases. But it doesn't really let us build stuff on top of those methods once, without "infectious codegen". An example is the new `fromCreatedEvent`: this method is now hand-written, with only static forwarding shims codegenned.

Other things can usefully be moved into `ContractCompanion` and its signature and hierarchy usefully expanded. `fromValue` is a really good candidate. But this PR is doing quite enough.

```rst
CHANGELOG_BEGIN
- [daml codegen js] Generated templates have types, inherited methods,
  and a new static ``COMPANION`` that enable writing more
  template-generic utilities directly in Java.  New classes
  ``Contract``, ``ContractWithKey``, and ``ContractCompanion``, have
  been introduced; if you are importing
  ``com.daml.ledger.javaapi.data.codegen.*`` and have imported classes
  with these names from any other package, these references will stop
  working; use individual imports instead.
  See `issue #13724 <https://github.com/digital-asset/daml/pull/13724>`__.
CHANGELOG_END
```

* [x] keyed contracts
* [x] companion
* [x] eliminate generated contract toString?
* [x] test abstract usage
* [x] changelog
* [x] remove abstract deprecated method, not needed for static forwarders
* [ ] make `Contract` vs `ContractWithKey` more like `ContractCompanion.WithoutKey` vs `ContractCompanion.WithKey`?
* [ ] call `ContractCompanion` `TemplateCompanion` instead, as with Scala codegen?

### CODE REVIEW

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+public abstract class ContractCompanion<Ct, Id, Data> {
+  public final Identifier TEMPLATE_ID;
+  final String templateClassName; // not something we want outside this package
+
+  protected final Function<String, Id> newContractId;
+  protected final Function<DamlRecord, Data> fromValue;
+
+  @Deprecated
+  public abstract Ct fromIdAndRecord(String contractId, DamlRecord record$);
```
Comment (ray-roestenburg-da) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java:23]: Would it make sense not to add an immediately deprecated method?

**Review by ray-roestenburg-da — APPROVED** [review]

**Review by stefanobaghino-da — DISMISSED** [review]
Adding a blocker to make sure a few important points don't go unaddressed:
* If a client changes a Daml model, upgrades to a new version where this PR got merged and re-generates code purely with the intent of generating code for the changed Daml models, will existing client code cease to compile, raise compilation warnings or change runtime behavior?
* How does this feature interact with the decoder class? Is there a change in approach enabled by this change?
* This adds a new feature and/or changes the way in which an existing feature can be changed: is this sufficiently documented to enable users to start using this independently after going through the documentation?

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java (hunk 1 above)
Comment (S11001001) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java:23]: We want the new concrete methods to be deprecated, as they are only there to support existing deprecated methods that are now static forwarders (removing all of these can be a separate issue), but we can

* [ ] remove the deprecated abstract parent

**Comment by S11001001** [comment]
> If a client changes a Daml model, upgrades to a new version where this PR got merged and re-generates code purely with the intent of generating code for the changed Daml models, will existing client code cease to compile,
 
 @stefanobaghino-da Their code may fail `javac` if they `import com.daml.ledger.javaapi.data.codegen.*` _and also_ `com.daml.ledger.javaapi.data.Contract` (wildcard or no), and refer to `Contract`. No warnings or behavior changes.
 
 > raise compilation warnings or change runtime behavior?
 
 No.
 
 > How does this feature interact with the decoder class?
 
 Not in this PR.
 
 > Is there a change in approach enabled by this change?
 
 Yes, discussed in #13768.
 
 > This adds a new feature and/or changes the way in which an existing feature can be changed: is this sufficiently documented to enable users to start using this independently after going through the documentation?
 
 A user who is motivated to build the kinds of template-generic functions this PR enables should find all they need to know in the type signatures, with the understanding that they are just like the static methods of the same signature.
 
 If we were happy with the documentation of `fromCreatedEvent` before, then this feature offers exactly the same experience. If we want to expand on that, the Javadoc is the best place to do so; I do not think docs.daml.com should cover this in detail other than to mention that it is generated, as the example here does.

**Comment by stefanobaghino-da** [comment]
> @stefanobaghino-da Their code may fail javac if they import com.daml.ledger.javaapi.data.codegen.* and also com.daml.ledger.javaapi.data.Contract (wildcard or no), and refer to Contract. No warnings or behavior changes.

Please add this consideration to the change log so that the release coordinator can be aware of this. It's very important that we minimize the impact of changes and give a clear explanation on how to migrate if there is an impact. I'll remove my blocker in light of your replies. Thanks.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java (hunk 1 above)
Comment (S11001001) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java]: 0de313ed55743a0e0c9b980b627450d314aaa5a2

**Pull request merged**